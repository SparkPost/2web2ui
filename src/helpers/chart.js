import moment from 'moment';
import _ from 'lodash';
import { getPrecisionType } from './metrics';
import { roundToPlaces } from 'src/helpers/units';

function getDayLines(data, precision = 'day') {
  if (getPrecisionType(precision) !== 'hour') {
    return [];
  }

  const lastIndex = data.length - 1;
  return data.filter(({ ts }, i) => {
    if (i === 0 || i === lastIndex) {
      return false;
    }
    if (moment(ts).hours() !== 0) {
      return false;
    }
    return true;
  });
}

const getTimeTickFormatter = _.memoize((precisionType, { includeTimezone } = {}) => {
  const formatFunction = includeTimezone ? moment.parseZone : moment;
  const format = tickFormat => tick => formatFunction(tick).format(tickFormat);

  switch (precisionType) {
    case 'hour':
      return format('h:mma');

    case 'day':
    case 'week':
      return format('MMM Do');

    case 'month':
      return format('MMM');

    default:
      return format('MMM Do');
  }
});

const getTooltipLabelFormatter = _.memoize((precisionType, { includeTimezone } = {}) => {
  const formatFunction = includeTimezone ? moment.parseZone : moment;
  const format = labelFormat => label => formatFunction(label).format(labelFormat);
  switch (precisionType) {
    case 'hour':
      return format('MMM Do [at] LT');

    case 'day':
      return format('MMMM Do');

    case 'month':
      return format('MMMM YYYY');

    default:
      return format('MMMM Do');
  }
});

const getWeekPrecisionFormatter = to => {
  return label => {
    const format = momentTime => momentTime.format('MMM Do');
    const startDate = format(moment.parseZone(label));
    //If it's the last date, make the end date the to date; else, make the to date the next Saturday
    if (moment.parseZone(label).isSame(moment.parseZone(to).weekday(0), 'day')) {
      return `${startDate} - ${format(moment.parseZone(to))}`;
    }

    return `${startDate} - ${format(moment.parseZone(label).weekday(6))}`;
  };
};

function getLineChartFormatters(precision, to = moment(), { includeTimezone } = {}) {
  const formatters = {};
  const precisionType = getPrecisionType(precision);

  formatters.xTickFormatter = getTimeTickFormatter(precisionType, { includeTimezone });
  formatters.tooltipLabelFormatter =
    precisionType === 'week' //Can't put in case;switch in getToolTipLabelFormatter b/c memoization
      ? getWeekPrecisionFormatter(to)
      : getTooltipLabelFormatter(precisionType, { includeTimezone });

  return formatters;
}

function cumulativeSum({ data, key }) {
  const results = data.reduce((cumulativeDataSet, data) => {
    const cumulativeData = { ...data };

    if (cumulativeDataSet.length > 0 && data[key]) {
      cumulativeData[key] = data[key] + cumulativeDataSet[cumulativeDataSet.length - 1][key];
    }

    cumulativeDataSet = [...cumulativeDataSet, cumulativeData];

    return cumulativeDataSet;
  }, []);

  return results;
}

const formatYAxisPercent = _.memoize(v => `${roundToPlaces(v, v < 1 ? 3 : 1)}%`);

export {
  cumulativeSum,
  getDayLines,
  getTimeTickFormatter,
  getTooltipLabelFormatter,
  getWeekPrecisionFormatter,
  getLineChartFormatters,
  formatYAxisPercent,
};
