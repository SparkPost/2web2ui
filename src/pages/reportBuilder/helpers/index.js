import _ from 'lodash';
import {
  REPORT_BUILDER_FILTER_KEY_MAP,
  REPORT_BUILDER_FILTER_KEY_INVERTED_MAP,
} from 'src/constants';
import { COMPARE_BY_OPTIONS } from '../constants';

/**
 * @name getFilterTypeKey
 * @description Returns the filter key for REPORT_BUILDER_FILTER_KEY_MAP object based on the passed in value
 * This is needed for backwards compatibility
 * @param {string} key - either the key or label value for the filters map
 * @returns {string} returns human-readable filter type from the passed in string key, e.g. 'domains' returns 'Recipient Domain'
 */
export function getFilterTypeKey(key) {
  if (REPORT_BUILDER_FILTER_KEY_INVERTED_MAP.hasOwnProperty(key)) {
    return key;
  }

  return REPORT_BUILDER_FILTER_KEY_MAP[key];
}

/**
 * @name getFilterTypeLabel
 * @description Returns the filter label for REPORT_BUILDER_FILTER_KEY_MAP object based on the passed in value
 * @param {string} key either the key or label value for the filters map
 * @returns {string} the relevant, human-readable filter label
 */
export function getFilterTypeLabel(key) {
  if (REPORT_BUILDER_FILTER_KEY_MAP.hasOwnProperty(key)) {
    if (key === 'Campaign') {
      const campaignKey = REPORT_BUILDER_FILTER_KEY_MAP[key];
      return REPORT_BUILDER_FILTER_KEY_INVERTED_MAP[campaignKey];
    }

    return key;
  }

  return REPORT_BUILDER_FILTER_KEY_INVERTED_MAP[key];
}
/**
 * @name getIterableFormattedGroupings
 * @description remaps advanced filter groupings data to make it iterable for use by the UI
 * @param {Array[Object]} groupings - An array of grouped [advanced filters](https://developers.sparkpost.com/api/metrics/#header-advanced-filters)
 * @returns {Array[Object]} returns an array of UI-consumable filters
 * */
export function getIterableFormattedGroupings(groupings) {
  return groupings.map(grouping => {
    const groupingKeys = Object.keys(grouping);
    const groupingType = groupingKeys[0];
    const filters = Object.keys(grouping[groupingType]);

    if (filters.length === 0) {
      return {
        type: groupingType,
        filters: [getInitialFilterState()],
      };
    }

    return {
      type: groupingType,
      filters: _.flatten(
        filters.map(filter => {
          const filterObj = grouping[groupingType][filter];
          const comparisons = Object.keys(filterObj);

          return comparisons.map(compareBy => {
            return {
              type: filter,
              compareBy: compareBy,
              values: _.get(filterObj, compareBy),
            };
          });
        }),
      ),
    };
  });
}

/**
 * @name getInitialFilterState
 * @description returns initial state for an individual filter, i.e., after it is added to a list of filters
 * @returns {Object} UI-consumable, initial, filter state
 */
export function getInitialFilterState() {
  return {
    type: undefined,
    compareBy: 'eq',
    values: [],
  };
}

/**
 * @name getInitialGroupState
 * @description returns initial state for a group, i.e., after it is added to a list of groupings
 * @returns {Object} UI-consumable, initial filter group state
 */
export function getInitialGroupState() {
  return {
    type: 'AND',
    filters: [getInitialFilterState()],
  };
}

/**
 * @name getGroupingFields
 * @description returns UI state to render multi-layer filter forms based on the user's entry
 * Unit tests for this function were intentionally not written as they are tied closely to UI state
 * and are well tested by integration tests.
 * @param {Array} iterableGroupings array of iterable filter groupings, usually derived from `getIterableFormattedGroupings`
 * @returns {Array[Object]} UI-consumable array of objects that describe the state of a series of filter group form fields. Used within the Analytics Report filters form UI.
 */
export function getGroupingFields(iterableGroupings) {
  const groupings = iterableGroupings;
  const getValueFieldType = compareBy => {
    switch (compareBy) {
      case 'eq':
      case 'notEq':
        return 'typeahead';

      case 'like':
      case 'notLike':
        return 'multi-entry';

      default:
        return undefined;
    }
  };

  return groupings.map((grouping, groupingIndex) => {
    return {
      ...grouping,

      // Renders an "AND" below all groups *except* the last one in the list
      hasAndBetweenGroups: Boolean(groupingIndex + 1 < groupings.length),

      // Renders below the last group in the list when one of the filters has values
      hasAndButton: Boolean(
        groupingIndex + 1 === groupings.length &&
          groupings[groupingIndex].filters.find(filter => filter.values.length > 0),
      ),

      // Whether or not this particular group has duplicate filters
      hasDuplicateFilters: getHasDuplicateFilters(grouping.filters),

      filters: grouping.filters.map((filter, filterIndex) => {
        return {
          ...filter,

          // Grabbing the label by the key value
          label: Object.keys(REPORT_BUILDER_FILTER_KEY_MAP).find(
            key => REPORT_BUILDER_FILTER_KEY_MAP[key] === filter.type,
          ),

          // The form comparison field does not render without a type selected,
          hasCompareBySelect: !!filter.type,

          // Controls whether a filter has "like" comparison options
          hasCompareByLikeOptions: !!filter.type && filter.type !== 'subaccounts',

          // Filter comparison type controls which type of form control renders next
          valueField: filter.type ? getValueFieldType(filter.compareBy) : undefined,

          // Renders on the first group of filters only and when valid filter values exist
          hasGroupingTypeRadioGroup: filterIndex === 0 && filter.values.length > 0,

          // Renders when the grouping is an "AND", when valid filter values exist, on the last filter
          hasAndButton:
            grouping.type === 'AND' &&
            filter.values.length > 0 &&
            filterIndex + 1 === grouping.filters.length,

          // Renders when the grouping is an "OR", when valid filter values exist, on the last filter
          hasOrButton:
            grouping.type === 'OR' &&
            filter.values.length &&
            filterIndex + 1 === grouping.filters.length,

          // Renders comparison text ("OR" or "AND") when more than one grouping exists but not on the last filter in the grouping
          hasComparisonBetweenFilters:
            grouping.filters.length > 1 && filterIndex + 1 !== grouping.filters.length,

          // Renders when there is more than 1 grouping in the form
          hasRemoveButton:
            (groupingIndex === 0 && grouping.filters.length > 1) || groupingIndex > 0,
        };
      }),
    };
  });
}

/**
 * @name getActiveFilterTagGroups
 * @description returns UI state to render active filter tags based on the user's entry
 * Unit tests for this function were intentionally not written as they are tied closely to UI state
 * and are well tested by integration tests.
 * @param {Array} iterableGroupings array of iterable filter groupings, usually derived from `getIterableFormattedGroupings`
 * @returns {Array[Object]} UI-consumable array of objects that describes the state of active filter tags rendered to the page
 */
export function getActiveFilterTagGroups(iterableGroupings) {
  const groupings = iterableGroupings;
  const getCompareByText = compareBy => {
    switch (compareBy) {
      case 'eq':
        return 'is equal to';
      case 'notEq':
        return 'is not equal to';
      case 'like':
        return 'contains';
      case 'notLike':
        return 'does not contain';
      default:
        throw new Error(`${compareBy} is not a valid comparison value.`);
    }
  };

  return groupings.map((grouping, groupingIndex) => {
    return {
      type: grouping.type,
      values: grouping.values,

      // Renders below all groups *except* the last one in the list
      hasAndBetweenGroups: groupingIndex + 1 < groupings.length,

      filters: grouping.filters.map((filter, filterIndex) => {
        return {
          values: filter.values,

          // Grabbing the label by the key value
          label: REPORT_BUILDER_FILTER_KEY_INVERTED_MAP[filter.type],

          compareBy: getCompareByText(filter.compareBy),

          // Renders comparison text ("OR" or "AND") when more than one grouping exists but not on the last filter in the grouping
          hasComparisonBetweenFilters:
            grouping.filters.length > 1 && filterIndex + 1 !== grouping.filters.length,
        };
      }),
    };
  });
}

/**
 * @name getApiFormattedGroupings
 * @description reverts UI-formatted data mapping and re-structures said data to match API structure
 * @param {Array[Object]} groupingFields - array of UI filter group fields, typically derived from `getGroupingFields`
 * @returns {Array[Object]} array of API-consumable [advanced filters](https://developers.sparkpost.com/api/metrics/#header-advanced-filters)
 */
export function getApiFormattedGroupings(groupingFields) {
  return groupingFields
    .map(grouping => {
      const groupingHasValues = Boolean(grouping.filters.find(filter => filter.values.length > 0));
      const groupingType = grouping.type; // "AND" or "OR"

      // Only reformat and return a grouping when it has valid filters within
      if (!groupingHasValues) return undefined;

      // Converts filters array to object/key structure
      const formattedFilters = grouping.filters.reduce((obj, filter) => {
        const filterHasValues = filter.values.length > 0;
        const comparisonWithValues = _.setWith({}, filter.compareBy, filter.values); // Pairs values with the grouping type, i.e., `"domains": [ ...values ]`
        const filterObj = {
          [filter.type]: {
            ...obj[filter.type],
            ...comparisonWithValues,
          },
        };

        return {
          ...obj,
          ...(filterHasValues ? filterObj : undefined), // Only incorporate the object key and its comparisons when values are present
        };
      }, {});

      return {
        [groupingType]: {
          ...formattedFilters,
        },
      };
    })
    .filter(Boolean); // Remove undefined entries in the object
}

/**
 * @name hydrateFilters
 * @description Converts API data structure in to a more human-readable version of filters for later rendering
 * @param {Array} groupings grouped [advanced filters](https://developers.sparkpost.com/api/metrics/#header-advanced-filters) retrieved from the API
 * @param {Object} options
 *   @param {Array} options.subaccounts array of subaccounts, usually retrieved from the global data store
 * @returns UI-consumable filters for rendering the currently active filters within a component
 */
export function hydrateFilters(groupings, { subaccounts } = {}) {
  return groupings.map(grouping => {
    const groupingKeys = Object.keys(grouping);
    const groupingType = groupingKeys[0];
    const filters = Object.keys(grouping[groupingType]);

    return {
      [groupingType]: filters.reduce((ret, filter) => {
        const filterObj = grouping[groupingType][filter];
        const comparisons = Object.keys(filterObj);
        ret[filter] = comparisons.reduce((filterRet, comparison) => {
          switch (comparison) {
            case 'eq':
            case 'notEq': {
              if (filter === 'subaccounts') {
                filterRet[comparison] = filterObj[comparison].map(rawValue => {
                  // Depending on the filters being parsed (old vs. newer grouped comparator filters vs custom),
                  // the filters may be a an object instead of a string.
                  const id = typeof rawValue === 'object' ? rawValue.id : rawValue;
                  const subaccount = subaccounts.find(
                    subaccount => subaccount.id === Number.parseInt(id),
                  );

                  return {
                    value: subaccount ? `${subaccount.name} (ID ${id})` : id,
                    id,
                    type: getFilterTypeKey(filter),
                  };
                });
              } else {
                filterRet[comparison] = filterObj[comparison].map(rawValue => {
                  // Depending on the filters being parsed (old vs. newer grouped comparator filters),
                  // the filters may be a an object instead of a string.
                  const value = typeof rawValue === 'object' ? rawValue.value : rawValue;

                  return { value, type: getFilterTypeKey(filter) };
                });
              }

              return filterRet;
            }
            case 'like':
            case 'notLike':
            default:
              filterRet[comparison] = filterObj[comparison];
              return filterRet;
          }
        }, {});

        return ret;
      }, {}),
    };
  });
}

/**
 * @name dehydrateFilters
 * @description flattens manipulated data structure for use with the API
 * @param {Array} groupings UI-consumable filters as returned by `hydrateFilters`
 * @returns {Array[Object]} array of API-consumable [advanced filters](https://developers.sparkpost.com/api/metrics/#header-advanced-filters)
 */
export function dehydrateFilters(groupings) {
  return groupings.map(grouping => {
    const groupingKeys = Object.keys(grouping);
    const groupingType = groupingKeys[0];
    const filters = Object.keys(grouping[groupingType]);

    return {
      [groupingType]: filters.reduce((ret, filter) => {
        const filterObj = grouping[groupingType][filter];
        const comparisons = Object.keys(filterObj);
        ret[filter] = comparisons.reduce((filterRet, comparison) => {
          switch (comparison) {
            case 'eq':
            case 'notEq': {
              if (filter === 'subaccounts') {
                filterRet[comparison] = filterObj[comparison].map(({ id }) => String(id));
              } else {
                filterRet[comparison] = filterObj[comparison].map(({ value }) => value);
              }

              return filterRet;
            }
            case 'like':
            case 'notLike':
            default:
              filterRet[comparison] = filterObj[comparison];
              return filterRet;
          }
        }, {});

        return ret;
      }, {}),
    };
  });
}

/**
 * @name getHasDuplicateFilters
 * @description within a filter grouping, this function returns `true` when duplicate filters are present.
 * Only the "type" and "compareBy" keys are relevant when making this comparison.
 * @param {Array[Object]} filters - iterable array of filter objects derived from `getIterableFormattedGroupings`
 * @returns {boolean} whether or not the group of filters has a duplicate
 */
export function getHasDuplicateFilters(filters) {
  // Simplifies filters according to relevant keys
  const formattedFilters = filters.map(({ type, compareBy }) => {
    return {
      type,
      compareBy,
    };
  });
  const uniqueFilters = _.uniqWith(formattedFilters, _.isEqual); // Generates an array of unique filters

  // If there are more filters than unique filters, then there must be a duplicate present
  return filters.length > uniqueFilters.length;
}

/**
 * Hashmap of valid filter types
 */
const VALID_FILTERS_KEY_MAP = Object.values(REPORT_BUILDER_FILTER_KEY_MAP || {}).reduce(
  (map, key) => {
    //The `|| {}` is needed to fix some breaking unit tests.
    return map.set(key, true);
  },
  new Map(),
);
/**
 * Hashmap of valid comparator values
 */
const VALID_COMPARE_BY_LOWERCASE_MAP = COMPARE_BY_OPTIONS.reduce(
  (accumulator, { value }) => accumulator.set(value.toLowerCase(), true),
  new Map(),
);
/**
 * @name getValidFilters
 * @description returns a new array of only valid filters. It only removes at the top level. If any part of
 * the filter is invalid, it will remove the entire top level filter.
 * @param {Array[Object]} filters - iterable array of filter objects derived from `hydrateFilters`
 * @returns {Array[Object]} array of API-consumable [advanced filters](https://developers.sparkpost.com/api/metrics/#header-advanced-filters)
 */
export function getValidFilters(filters) {
  // Remaps top level comparators to true/false
  // EX: [{AND:{...}},{OR:{...}}] => [true ,false] where true = valid filter and false = invalid filter
  const validFiltersArray = filters.map(grouping => {
    // Checks that every comparator is formatted correctly
    if (typeof grouping !== 'object' || Object.keys(grouping).length < 1) {
      return false;
    }
    const groupingType = Object.keys(grouping)[0];
    // Checks that the top level comparison conjugate is AND/OR
    if (groupingType.toUpperCase() !== 'AND' && groupingType.toUpperCase() !== 'OR') {
      return false;
    }

    // Checks that every filter is formatted correctly
    if (
      typeof grouping[groupingType] !== 'object' ||
      Object.keys(grouping[groupingType]).length < 1
    ) {
      return false;
    }

    // Checks that every filter is valid
    const filterType = grouping[groupingType];
    return Object.keys(filterType).every(filter => {
      if (!VALID_FILTERS_KEY_MAP.has(filter) || typeof filterType[filter] !== 'object') {
        return false;
      }
      const comparators = filterType[filter];
      // Checks that every comparator is allowed
      return Object.keys(comparators).every(comparator => {
        return (
          VALID_COMPARE_BY_LOWERCASE_MAP.has(comparator.toLowerCase()) &&
          Array.isArray(comparators[comparator])
        );
      });
    });
  });

  // Re-maps the truthy filters to the actual filters and removes all the falsy ones.
  return validFiltersArray
    .map((isValidFilter, index) => (isValidFilter ? filters[index] : false))
    .filter(Boolean);
}

/**
 * @name replaceComparisonFilterKey
 * @description replaces the key of comparisons with the key value rather than label value
 * @param {Array[Object]} comparisons array of comparison objects derived from the UI "Add Comparison" form
 * @returns {Array[Object]} returns re-formatted array of comparisons
 */
export function replaceComparisonFilterKey(comparisons) {
  return comparisons
    .map(comparison => {
      return { ...comparison, type: getFilterTypeKey(comparison.type) };
    })
    .filter(({ type }) => Boolean(type));
}
