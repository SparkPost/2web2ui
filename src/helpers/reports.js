import _ from 'lodash';
import qs from 'qs';
import { getRelativeDates, relativeDateOptions } from 'src/helpers/date';
import { stringifyTypeaheadfilter } from 'src/helpers/string';
import { REPORT_BUILDER_FILTER_KEY_MAP } from 'src/constants';
import { list as METRICS_LIST } from 'src/config/metrics';
export function dedupeFilters(filters) {
  return _.uniqBy(filters, stringifyTypeaheadfilter);
}

/**
 * Parses search string
 * @param  {string} search - location.search
 * @returns {Object}
 * {
 *   filters: array of advanced filters
 *   comparisons: array of filters that are being compared
 *   range: specific range for dates, used in place of specific dates
 *   from: javascript date representing beginning of range
 *   to: javascript date representing beginning of range
 *   timezone: string representing timezone
 *   precision: string representing grouping of data based on a time intervals (minute, hour, day, etc.)
 *  report: string representing ID of an active report
 * }
 */
export function parseSearch(search) {
  if (!search) {
    return {};
  }

  const {
    from,
    to,
    range,
    metrics,
    timezone,
    precision,
    filters = [],
    query_filters,
    comparisons,
    report,
    industryBenchmarkMetric,
    industryBenchmarkFilters,
  } = qs.parse(search, { ignoreQueryPrefix: true, arrayLimit: METRICS_LIST.length });
  let ret = {};

  if (report) {
    ret.report = report;
  }

  if (query_filters) {
    try {
      ret.filters = JSON.parse(decodeURI(query_filters));
    } catch {
      ret.filters = [];
    }
  } else {
    const filtersList = (typeof filters === 'string' ? [filters] : filters).map(filter => {
      const parts = filter.split(':');
      const type = parts.shift();
      let value;
      let id;

      // Subaccount filters include 3 parts
      // 'Subaccount:1234 (ID 554):554' -> { type: 'Subaccount', value: '1234 (ID 554)', id: '554' }
      if (type === 'Subaccount') {
        id = parts.pop();
        value = parts.join(':');
      } else {
        value = parts.join(':');
      }
      return { value, type, id };
    });

    ret.filters = mapFiltersToComparators(filtersList);
  }

  if (comparisons) {
    ret.comparisons = comparisons;
  }

  if (metrics) {
    ret.metrics = typeof metrics === 'string' ? [metrics] : metrics;
  }

  const fromTime = new Date(from);
  const toTime = new Date(to);

  if (from && !isNaN(fromTime)) {
    ret.from = fromTime;
  }

  if (to && !isNaN(toTime)) {
    ret.to = toTime;
  }

  if (range) {
    const invalidRange = !_.find(relativeDateOptions, ['value', range]);
    const effectiveRange = invalidRange ? 'day' : range;
    ret = { ...ret, ...getRelativeDates(effectiveRange) };
  }

  if (precision) {
    ret.precision = precision;
  }

  if (timezone) {
    ret.timezone = timezone;
  }

  if (industryBenchmarkMetric) {
    ret.industryBenchmarkMetric = industryBenchmarkMetric;
    if (industryBenchmarkFilters) {
      try {
        ret.industryBenchmarkFilters = JSON.parse(decodeURI(industryBenchmarkFilters));
      } catch {
        ret.industryBenchmarkFilters = { industryCategory: 'all', mailboxProvider: 'all' };
      }
    }
  }

  // filters are used in pages to dispatch updates to Redux store
  return ret;
}

export function mapFiltersToComparators(filters) {
  const mappedFilters = filters.reduce((acc, { value, type, id }) => {
    const filterKey = REPORT_BUILDER_FILTER_KEY_MAP[type];
    acc[filterKey] = acc[filterKey] || { eq: [] };
    acc[filterKey].eq.push(filterKey === 'subaccounts' ? id : value);
    return acc;
  }, {});

  if (_.isEmpty(mappedFilters)) return [];

  return [{ AND: mappedFilters }];
}
