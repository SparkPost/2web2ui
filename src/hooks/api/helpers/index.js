import { useDispatch, useSelector } from 'react-redux';
import { useQueryClient } from 'react-query';
import { refresh, logout } from 'src/actions/auth';
import { showAlert } from 'src/actions/globalAlert';
import { fetch as fetchAccount } from 'src/actions/account';
import { useRefreshToken } from 'src/helpers/http';

/**
 * Common preparation steps composed of other hooks
 */
export function usePrepareQuery() {
  const queryClient = useQueryClient();
  const auth = useSelector(state => state.auth);
  const dispatch = useDispatch();

  return {
    queryClient,
    auth,
    dispatch,
  };
}

/**
 * Handle errors based on the response from a failed query
 *
 * @param {object} error - error returned from a failed API request
 * @param {string} method - HTTP request method
 * @param {object} queryClient - `react-query` queryClient - https://react-query.tanstack.com/reference/QueryClient
 * @param {object} auth - auth state stored in global Redux store
 * @param {function} dispatch - `react-redux` dispatch method, typically derived from `useDispatch()`
 *
 */
export function handleError({ error, method, queryClient, auth, dispatch }) {
  const { response = {} } = error;

  if (response.status === 401 && auth.refreshToken) {
    // Invalidate any in-progress queries
    queryClient.invalidateQueries();

    return (
      // This isn't actually a hook ðŸ¤”
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useRefreshToken(auth.refreshToken) // eslint-disable-line react-hooks/rules-of-hooks
        // dispatch a refresh action to save new token results in cookie and store
        .then(({ data } = {}) => dispatch(refresh(data.access_token, auth.refreshToken)))

        // dispatch the original action again, now that we have a new token ...
        // if anything in this refresh flow blew up, log out
        .then(
          // refresh token request succeeded
          async () => {
            // Refetch queries when the refresh token is successful
            // See: https://react-query.tanstack.com/docs/api/#queryClientrefetchqueries
            return await queryClient.refetchQueries();
          },
          // refresh token request failed
          err => {
            dispatch(logout());
            throw err;
          },
        )
    );
  }

  if (response.status === 401) {
    return dispatch(logout());
  }

  if (response.status === 403) {
    return dispatch(fetchAccount());
  }

  // No messaging necessary for 404 errors for "GET" requests
  if (response.status === 404 && method === 'GET') {
    return;
  }

  // This error handling could be handled on the server, however, the effort required to do so was significantly larger than handling it here
  if (response.status === 413) {
    return dispatch(
      showAlert({
        type: 'error',
        message: 'Something went wrong.',
        details: 'File size larger than the server allows.',
      }),
    );
  }

  return dispatch(
    showAlert({
      type: 'error',
      message: 'Something went wrong.',
      details: error.message,
    }),
  );
}

/**
 * Derive query key from passed in queryFn and auth state from Redux store. This would have been a custom hook, however, needed to be run within a callback - impossible with a hook.
 * See: https://react-query.tanstack.com/docs/guides/queries#query-keys
 *
 * @param {function} queryFn - query function derived from `src/helpers/api` helpers
 * @param {object} auth - auth Redux state
 *
 */
export function deriveQueryKey({ queryFn, auth }) {
  const { url, method, params, headers, data } = queryFn();

  return [url, { method, params, headers, auth, data }];
}
