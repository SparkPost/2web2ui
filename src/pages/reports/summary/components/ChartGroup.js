import React, { Component } from 'react';
import _ from 'lodash';
import { getDayLines, getLineChartFormatters } from 'src/helpers/chart';
import LineChart from './LineChart'; // async
import METRICS_UNIT_CONFIG from 'src/config/metrics-units';
const DEFAULT_UNIT = 'number';

function getUniqueUnits(metrics) {
  return _.uniq(metrics.map(({ unit = DEFAULT_UNIT }) => unit));
}

export default class ChartGroup extends Component {
  createDayReferenceLines() {
    const { chartData, precision } = this.props;

    return getDayLines(chartData, precision).map(({ ts }) => ({
      key: ts,
      x: ts,
      stroke: '#bbb',
      strokeWidth: 2,
    }));
  }

  render() {
    const { chartData = [], metrics, chartLoading, precision, yScale, to } = this.props;

    if (!chartData.length || !metrics) {
      return null;
    }

    const formatters = getLineChartFormatters(precision, to);
    const referenceLines = this.createDayReferenceLines();
    const charts = getUniqueUnits(metrics).map(unit => ({
      metrics: metrics.filter(metric => metric.unit === unit),
      ...METRICS_UNIT_CONFIG[unit],
    }));

    return (
      <div>
        {charts.map((chart, i) => (
          <LineChart
            key={`chart=${i}`}
            syncId="summaryChart"
            data={chartData}
            precision={precision}
            lines={chart.metrics.map(({ name, label, stroke }) => ({
              key: name,
              dataKey: name,
              name: label,
              stroke: chartLoading ? '#f8f8f8' : stroke,
            }))}
            {...formatters}
            yTickFormatter={chart.yAxisFormatter}
            yScale={yScale}
            yLabel={chart.label}
            tooltipValueFormatter={chart.yAxisFormatter}
            referenceLines={referenceLines}
            showXAxis={i === charts.length - 1}
          />
        ))}
      </div>
    );
  }
}
