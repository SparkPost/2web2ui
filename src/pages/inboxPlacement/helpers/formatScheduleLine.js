import React from 'react';
import moment from 'moment';
import { STATUS } from '../constants/test';
import { FORMATS } from 'src/constants';
import { InlineItems } from 'src/components/structure';
import { Bold } from 'src/components/text';

const formatDateTime = start_time => moment(start_time).format(FORMATS.MOMENT.LONG_DATETIME_ALT);

export default function formatScheduleLine(status, start_time, end_time) {
  const formatted_start_time = formatDateTime(start_time);

  switch (status) {
    case STATUS.RUNNING: {
      return (
        <>
          <Bold>Sent</Bold> {formatted_start_time}
        </>
      );
    }
    case STATUS.STOPPED: {
      return (
        <InlineItems compact>
          <span>
            <Bold>Sent</Bold> {formatted_start_time}
          </span>
          <span>
            <Bold>Stopped</Bold> {formatDateTime(end_time)}
          </span>
        </InlineItems>
      );
    }
    case STATUS.COMPLETED: {
      return (
        <InlineItems compact>
          <span>
            <Bold>Sent</Bold> {formatted_start_time}
          </span>
          <span>
            <Bold>Completed</Bold> {formatDateTime(end_time)}
          </span>
        </InlineItems>
      );
    }
    default: {
      return (
        <>
          <Bold>Sent</Bold> {formatted_start_time} and test status unknown
        </>
      );
    }
  }
}
