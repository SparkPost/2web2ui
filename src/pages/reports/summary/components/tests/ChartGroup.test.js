import React from 'react';
import { shallow } from 'enzyme';
import ChartGroup from '../ChartGroup';
import * as chartHelpers from 'src/helpers/chart';

jest.mock('src/helpers/chart');

describe('Component: Summary Chart Group', () => {

  let wrapper;
  let props;

  beforeEach(() => {
    chartHelpers.getLineChartFormatters = jest.fn(() => ({ formatter: 'mock' }));
    chartHelpers.getDayLines = jest.fn(() => [{ ts: 'day lines mock' }]);
    props = {
      chartData: [1, 2, 3],
      precision: 'day',
      metrics: [
        { name: 'a', label: 'a', stroke: 'a' }
      ],
      chartLoading: false,
      yScale: 'linear'
    };
    wrapper = shallow(<ChartGroup {...props} />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render when loading', () => {
    wrapper.setProps({ chartLoading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render null if no chart data', () => {
    wrapper.setProps({ chartData: []});
    expect(wrapper).toMatchSnapshot();
  });

  it('should render null if no metrics given', () => {
    wrapper.setProps({ metrics: undefined });
    expect(wrapper).toMatchSnapshot();
  });

});
