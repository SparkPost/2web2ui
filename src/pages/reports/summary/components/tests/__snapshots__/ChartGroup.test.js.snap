// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Summary Chart Group should render 1`] = `
<div>
  <LoadableComponent
    data={
      Array [
        1,
        2,
        3,
      ]
    }
    formatter="mock"
    key="chart=0"
    lines={Array []}
    precision="day"
    referenceLines={
      Array [
        Object {
          "key": "day lines mock",
          "stroke": "#bbb",
          "strokeWidth": 2,
          "x": "day lines mock",
        },
      ]
    }
    showXAxis={true}
    syncId="summaryChart"
    tooltipValueFormatter={[Function]}
    yLabel="Count"
    yScale="linear"
    yTickFormatter={[Function]}
  />
</div>
`;

exports[`Component: Summary Chart Group should render null if no chart data 1`] = `""`;

exports[`Component: Summary Chart Group should render null if no metrics given 1`] = `""`;

exports[`Component: Summary Chart Group should render when loading 1`] = `
<div>
  <LoadableComponent
    data={
      Array [
        1,
        2,
        3,
      ]
    }
    formatter="mock"
    key="chart=0"
    lines={Array []}
    precision="day"
    referenceLines={
      Array [
        Object {
          "key": "day lines mock",
          "stroke": "#bbb",
          "strokeWidth": 2,
          "x": "day lines mock",
        },
      ]
    }
    showXAxis={true}
    syncId="summaryChart"
    tooltipValueFormatter={[Function]}
    yLabel="Count"
    yScale="linear"
    yTickFormatter={[Function]}
  />
</div>
`;
